storage/migrations/01_oasis_3_consensus.up.sql

```diff
@@ -58,9 +58,7 @@ CREATE TABLE oasis_3.events
 
   txn_block  BIGINT NOT NULL,
   txn_hash   TEXT NOT NULL,
-  txn_index  INTEGER,
-
-  FOREIGN KEY (txn_block, txn_hash, txn_index) REFERENCES oasis_3.transactions(block, txn_hash, txn_index)
```

Comment:
> It would be preferable to instead just add DEFERRABLE INITIALLY DEFERRED at the end of this line, causing the constraint to be enforced at the end of each tx, rather than immediately when the row is inserted.

Agree with this principle. So, the problem is we are not currently indexing certain "transactions", and really neither are others. For example, this [event](https://www.oasisscan.com/events/10738156_2) has a transaction hash not [indexed](https://www.oasisscan.com/transactions/c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a).

Reply:
Discussed offline: The special transaction mentioned here (c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a) is the [SHA256 of the empty string](https://www.mathcelebrity.com/encryption.php?num=&pl=Encrypt), which is the event API's incredibly ugly way of saying "there is no associated tx". So we'll just map this to null, and make the db fields `txn_{block,hash,index}` nullable.
And it's OK to have null values even in composite FK ([example](https://dba.stackexchange.com/questions/265382/null-in-foreign-key-columns)).

Reply:
> make the db fields txn_{block,hash,index} nullable.

I think we only have to make `txn_hash` nullable. We still want the `block` `height` and the default `0` `index`.

Reply:
ah my bad, makes sense for the `height` to never be `null`.
doesn't make sense for the `index` to default to 0, unless I misunderstand it. `index==i` means "this event was produced in the `i`-th tx of block `height`", right?

Reply:
> index==i means "this event was produced in the i-th tx of block height", right?

Yeah, but I think it makes sense to remove that as well now. I don't think that is a meaningful reference on this table.

Reply:
Are you saying the entire foreign key is not meaningful? If so, I don't understand. Some events are still associated with a tx, right? And for those, we want to record the association.

Or are you saying only the index is not meaningful? If so, we do need it just as a technicality of the FK, which points to [this composite PK](https://github.com/oasisprotocol/oasis-indexer/blob/305095e8b1c04916924b517eeca4405d97586a90/storage/migrations/01_oasis_3_consensus.up.sql#L50-L50).

Reply:
> Are you saying the entire foreign key is not meaningful?

I think the foreign key is meaningful. Alternatively, we could do two indices (one to `blocks` table and another to `transactions`) but all that information is captured here in the composite FK.

> Or are you saying only the index is not meaningful? If so, we do need it just as a technicality of the FK, which points to [this composite PK](https://github.com/oasisprotocol/oasis-indexer/blob/305095e8b1c04916924b517eeca4405d97586a90/storage/migrations/01_oasis_3_consensus.up.sql#L50-L50).

Yeah, I don't think knowing which `txn_index` the particular `transaction` associated with the `event` is all that meaningful, and you could always do a subsequent lookup via the `transactions` table. My take is that we keep the foreign key and insert `NULL`.

Reply:
The definition of the composite PK (linked above) mentions why the index matters: the hash might not be unique.

Also, note that for composite FKs, if any single FK column is `NULL`, the entire FK constraint is lifted for that row.

So I think we do need to include the `txn_index`.

Reply:
Recap:

We agree to insert/include index when possible (aka for events associated with a transaction). For other events, they'll be null.